Improve form

mcritz opened the pull request:
Implements form `enctype` and `method` attributes. Includes tests.

This improvement makes the form API convenient by reducing code, and easier by using web standards without having to learn Plot API.

Before:
```
.form(
    .attribute(Attribute.attribute(named: "method", value: "POST")),
    .attribute(Attribute.attribute(named: "enctype", value: "multipart/form-data")),
    .input(
        .type(.submit)
    )
)
```

After:
```
.form(
    .method(.post),
    .enctype(.multipartFormData),
    .input(
        .type(.submit)
    )
)
```

kimar commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,36 @@
+    case applicationURLEncoded, multipartFormData, textPlain
+    
+    /// One of `applicationURLEncoded` (the default), `multipartFormData` (for use with files), or `textPlain` (conventionally used for debugging)
+    public var description: String {
```
Why not just use 

```swift
public enum HTMLFormEnctype: String, RawRepresentable {
    case applicationURLEncoded = "application/x-www-form-urlencoded"
    case multipartFormData = "multipart/form-data"
    case textPlain = "text/plain"
}
```

which results in a lot less code.
Also I wouldn't use the `description` property for reading the enum's value. https://developer.apple.com/documentation/swift/string/2431307-description

mcritz replied:
Updated to address feedback

kimar commented on Sources/Plot/API/HTMLAttributes.swift:
```diff
@@ -194,6 +194,18 @@ public extension Node where Context == HTML.FormContext {
+    /// Define the form’s Media Type.
+    /// - Parameter type: the Media Type the form will be serialized in.
+    static func enctype(_ type: HTMLFormEnctype) -> Node {
+        .attribute(named: "enctype", value: type.description)
```
With my proposed changes to `HTMLFormEnctype` this would be `type.rawValue` then.

mcritz replied:
Done

kimar commented on Sources/Plot/API/HTMLFormMethod.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    // Accodring the the HTML spec for form these forbidden `method`s should never be included: `CONNECT`, `TRACE`, and `TRACK`.
+    
+    public var description: String {
```
I would go with `RawRepresentable` (like I proposed in `HTMLFormEnctype `) here as well:

```swift
public enum HTMLFormMethod: String, RawRepresentable {
    // CORS-safelisted methods
    /// GET: By REST convention, read a resource.  Requests using GET should only retrieve data.
    case get = "GET"
    /// Like GET, but discared the body. Not a REST convetion.
    case head = "HEAD"
    /// By REST convention, create a resource.
    case post = "POST"
    
    // Acceptable methods
    /// DELETE: by REST convention, delete a resource.
    case delete = "DELETE"
    /// PUT: by REST convention, update a resource by *replacing* the existing resource with the request body.
    case put = "PUT"
    /// OPTIONS: used to describe the communication options for the target resource.
    case options = "OPTIONS"
    /// PATCH: by REST convention, update a resource by *modifing* part of the resouces with the request body.
    case patch = "PATCH"
}
```

Again I'm feeling like using the `description` property is code smell. Also I have the feeling explicitly defining the enums values (even though they currently are only the key's uppercased representation) would be beneficial, because: if there is a new key which would differ from the value then you're whole implementation might break.

mcritz replied:
Done

kimar commented on Sources/Plot/API/HTMLAttributes.swift:
```diff
@@ -194,6 +194,18 @@ public extension Node where Context == HTML.FormContext {
+    /// Defines the form’s HTTP method``
+    /// - Parameter method: HTTP verbs like `GET`, `POST`, `PUT`, `DELETE`, etc.
+    static func method(_ method: HTMLFormMethod) -> Node {
+        .attribute(named: "method", value: method.description)
```
`type.rawValue` if accepting the proposed changes to `HTMLFormMethod`.

kimar commented:
I definitely like this PR, really improves the form usage with Plot. 👍

JohnSundell commented:
Thanks a lot for this @mcritz (and thanks to @kimar for reviewing 😀), I'll take a look at this as soon as possible, but it most likely won't be until after New Year's Eve.

sindresorhus commented on Sources/Plot/API/HTMLAttributes.swift:
```diff
@@ -194,6 +194,18 @@ public extension Node where Context == HTML.FormContext {
+    
+    /// Define the form’s Media Type.
+    /// - Parameter type: the Media Type the form will be serialized in.
+    static func enctype(_ type: HTMLFormEnctype) -> Node {
```
Does the method name have to match the HTML attribute? As otherwise, I think it would be more readable with `encodingType()`.

mcritz replied:
I think that’s a topic of conversation. Let me know if you agree…

Choosing “enctype” matches HTML attributes. This is the right direction, otherwise we’re pretty much putting the HTML standards authorities on trial. I want to avoid something like, “You standards folks chose ‘enctype’, but we’re going to fix that here in our framework.” Ex: replace ‘enctype’ / ‘encodingType’ with ‘h1’ / ‘headerLevelOne’.

This also reduces the need to learn HTML and Plot’s “version” of HTML attributes.

JohnSundell replied:
Yeah having the names of these low-level node constructors match the names of their HTML equivalents is one of Plot's design goals (which we definitely need better docs on), for exactly the reasons that @mcritz mentions. Then we can build more higher-level components on top of those HTML primitives that use more "Swifty" naming conventions. For example, compare the `.stylesheet()` component to the underlying uses of `href()` and `rel()`.

mcritz commented:
@kimar Thanks for your feedback. I think it makes things cleaner and more aligned to the rest of the code in the codebase.

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+/// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#text/plain-encoding-algorithm
+public enum HTMLFormEnctype: String, RawRepresentable {
+    /// All characters are encoded. This is the default behavior of HTML <form> elements.
+    case applicationURLEncoded = "application/x-www-form-urlencoded"
```
Please see my comment on “plainText” below.

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+    case multipartFormData = "multipart/form-data"
+    
+    /// Spaces are converted to `+` character, but otherwise characters are not encoded. Used mostly for debugging.
+    case textPlain = "text/plain"
```
Why not just `plain` to match just `url`?

My naming convention for DOM attributes is meant to be internally consistent and approximately aligned to the HTML spec. Feel free to change this post-merge, if you’d like.

mcritz replied:
Here’s the type of naming convention I’m following: https://github.com/vapor/http/blob/ea25fb5c5b220de87df1fa44d4a5818615feb71a/Sources/HTTPKit/Headers/HTTPHeaders%2BName.swift#L33

mcritz commented on Tests/PlotTests/HTMLTests.swift:
```diff
@@ -249,7 +249,66 @@ final class HTMLTests: XCTestCase {
         """)
     }
+    
+    func testFormMultipartFormDataEncoded() {
```
Done

mcritz commented on Tests/PlotTests/HTMLTests.swift:
```diff
@@ -249,7 +249,66 @@ final class HTMLTests: XCTestCase {
+        </form></body>
+        """)
+        
+        html = HTML(.body(
```
Done

mcritz commented:
In the spirit of making things explicitly more searchable — and generalized! — I have renamed the types and files to accurately represent what they are and removed any comments that might have been helpful as Xcode option-clickable Quick Helps.

`HTMLFormEnctype` is now a generalized `MIMEType`. Which can now be generalized to include `image/jpeg` or `video/hevc`.
and
`HTMLFormMethod` is now a generalized `HTTPMethod` type.

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+//
```
Done

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+public enum HTMLFormEnctype: String, RawRepresentable {
+    /// All characters are encoded. This is the default behavior of HTML <form> elements.
+    case applicationURLEncoded = "application/x-www-form-urlencoded"
+    
```
Done

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+/// Describes the Media Type (also known as MIME type) of a form that defines the serialization of the form.
```
Refactored to be generalized MIMEType … er… type.

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+/// The actual encoding should be done by the browser / client. This enum merely describes the desired form behavior as HTML.
+/// See HTML spec 4.10.21.3
+/// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#text/plain-encoding-algorithm
+public enum HTMLFormEnctype: String, RawRepresentable {
```
Refactored to a generalized MIMEType

mcritz commented on Sources/Plot/API/HTMLFormEnctype.swift:
```diff
@@ -0,0 +1,23 @@
+    case applicationURLEncoded = "application/x-www-form-urlencoded"
+    
+    /// No characters are encoded. This value is required when you are using forms that have a file, media, images, video, upload control.
+    case multipartFormData = "multipart/form-data"
```
See my comment on “textPlain” above

JohnSundell commented:
Great! I'll take a second look at this, and will most likely merge it in (and fix any final small nitpicks myself post-merge) tomorrow 🙂 Thanks again @mcritz, and sorry that this has taken so long.

JohnSundell commented:
Post-merge fixes are here: https://github.com/JohnSundell/Plot/pull/23
